spec/lang/types.md

```diff
@@ -182,6 +196,40 @@ impl Type {
 And we also define how to compute the actual size and alignment.
 
 ```rust
+impl TupleHeadLayout {
+    /// The actual align of the tail, considering the packed attribute
+    fn capped_tail_align(self, tail_align: Align) -> Align {
+        match self.packed_align {
+            Some(packed_align) => tail_align.min(packed_align),
+            None => tail_align,
+        }
+    }
+
+    /// Where the tail starts, given the aligment of the tail type.
+    pub fn tail_offset(self, tail_align: Align) -> Offset {
+        let capped_tail_align = self.capped_tail_align(tail_align);
+        // `self.end.align_to(capped_tail_align)`.
+        let tail_offset = Size::from_bytes(self.end.bytes().next_multiple_of(capped_tail_align.bytes())).unwrap();
+        tail_offset
+    }
+
+    /// The size and alignment of the full tuple, including the tail.
+    /// Given the size and alignment of the tail type.
+    pub fn full_size_and_align(self, tail_size: Size, tail_align: Align) -> (Size, Align) {
+        let capped_tail_align = self.capped_tail_align(tail_align);
+        let align = capped_tail_align.max(self.align);
+        let tail_offset = self.tail_offset(tail_align);
+        let end = tail_offset + tail_size;
+        let size = Size::from_bytes(end.bytes().next_multiple_of(align.bytes())).unwrap();
+        (size, align)
+    }
+
+    /// Returns the size and alignment of the tuple when there is no tail.
+    pub fn head_size_and_align(self) -> (Size, Align) {
+        self.full_size_and_align(Size::ZERO, Align::ONE)
```

Comment:
Isn't this just `(self.end, self.align)`?

Reply:
No, because we align the end first. It is `self.end.align_to(self.align) though if you consider this nicer.